### PR TITLE
Ignore directories named pyproject.toml during settings and workspace discovery (Closes #14584)

### LINKED ISSUE #14584: uv commands would return `os error 21`
### Summary

Any `uv` commands that I executed always return:

```
error: failed to read from file `/pyproject.toml`: Is a directory (os error 21)
```

Adding `-v` to all commands does not return other helpful error messages.

What I have tried:
1. Remove cache
2. Uninstall (includes removing python and tool dirs) and reinstall
3. Remove cache, uninstall (including removing python and tool dirs), and reinstall

Could this be a duplicate issue? If yes, could you provide me the link? Appreciate the help.

Image

### Platform

Pop!_OS jammy 22.04 x86_64

### Version

uv 0.7.20

### Python version

Python 3.12

### PR DESCRIPTION
**Pull request by tyblue18**
## Summary

Any uv command ran with a *directory* named `pyproject.toml` (or `uv.toml`) anywhere between the current directory and the filesystem root crashed with a raw OS error:

```
error: failed to read from file `/pyproject.toml`: Is a directory (os error 21)
```

The crash site is settings discovery: `FilesystemOptions::find` walks up the directory tree during CLI config resolution, before any command logic runs, which is why every command was affected and also `from_directory` only tolerated `NotFound` when reading `uv.toml`/`pyproject.toml`.

Following (EliteTK) review, the fix now lives entirely in `uv-settings` (the earlier `uv-workspace` check has been reverted):

- `from_directory` returns a new `Error::Directory` when `uv.toml` or `pyproject.toml` is a directory, instead of just failing with the raw IO error.
- `FilesystemOptions::find` propagates that error IF and only IF it occurs at the top level of the search; for parent directories it emits a `warn!` (silent by default) and continues walking up.

Because this runs during config resolution, the top-level error now surfaces for *all* commands, including ones that don't do project discovery (`uv venv`, `uv pip ...`), and the same handling covers `uv.toml`.

One known edge: a `pyproject.toml` directory in the CWD is still silently ignored when a valid project exists in a parent directory. In that case the settings search starts at the workspace root and never visits the CWD. This seemed consistent with erroring only at the top level of the search, but happy to change this if a warning is preferred there instead.

## Before/after:

```
# directory named pyproject.toml (or uv.toml) in a parent of a valid project
before: error: failed to read from file `/tmp/repro/pyproject.toml`: Is a directory (os error 21)
after:  Resolved 1 package in 4ms
        (with -v: WARN Ignoring `/tmp/repro/pyproject.toml` during settings discovery: expected a file, found a directory)

# directory named pyproject.toml in the cwd
before: error: failed to read from file `/tmp/repro2/pyproject.toml`: Is a directory (os error 21)
after:  error: `pyproject.toml` is a directory, expected a file
```

Closes #14584

## Test Plan

Five integration tests using the standard snapshot harness:

- `lock_pyproject_toml_directory_in_parent` / `lock_uv_toml_directory_in_parent`: directory named `pyproject.toml`/`uv.toml` in a parent of a valid project → `uv lock` succeeds.
- `lock_pyproject_toml_directory_in_cwd`: directory named `pyproject.toml` in the cwd → clear error.
- `lock_uv_toml_directory_in_cwd`: directory named `uv.toml` in the cwd of an otherwise valid project → clear error.
- `create_venv_pyproject_toml_directory_in_cwd` (in `venv.rs`): the error also fires for commands that don't require a project.

Manually verified all scenarios with a built binary, including the `-v` warning, symlinks named `pyproject.toml` pointing at real files (still work, since `is_dir()` follows symlinks), and `uv venv --no-config` with a `pyproject.toml` directory in the cwd (still succeeds, as on main). Unit tests on `uv-settings`/`uv-workspace` pass; `cargo fmt --all --check` and `cargo clippy -p uv-workspace -p uv-settings --all-targets -- -D warnings` are clean.

### CODE REVIEW

File: crates/uv-settings/src/lib.rs [hunk 1]
```diff
@@ -189,6 +189,8 @@ impl FilesystemOptions {
                 return Ok(Some(Self(options)));
             }
             Err(err) if err.kind() == std::io::ErrorKind::NotFound => {}
+            // A directory named `pyproject.toml` is not a configuration file; skip it.
```
Comment (EliteTK) [inline crates/uv-settings/src/lib.rs]: The comment isn't needed.

There should be a warning here as discussed. I am torn between `warn_user_once!` and `warn!`. It could be that the directory is in a location the user cannot manage, so maybe err on the side of `warn!` (it's silent by default)?

While you're at it, we should also do the same for a `uv.toml` may also be a directory, in the block above.

File: crates/uv-workspace/src/workspace.rs [hunk 2]
```diff
@@ -264,6 +266,8 @@ impl Workspace {
             .clone();
         let path = normalize_path(&path);
 
+        check_pyproject_is_not_directory(&path)?;
+
```
Comment (EliteTK) [inline crates/uv-workspace/src/workspace.rs]: The discovery functions are the wrong place for this check, as they're not guaranteed to be only called on the CWD, and their errors are suppressed in a bunch of cases anyway.

Moreover, this is technically a breaking change.

**Comment by EliteTK** [comment]
The approach here isn't quite right.

With your PR we are actually losing a bunch of error cases when the current directory has a `pyproject.toml` directory.

The PR places the new error in the discovery path, which can end up called from a bunch of places and where the directory isn't guaranteed to be CWD.

We're also losing errors for the adjacent issue of `uv.toml` in CWD being a directory.

From the looks of it, the correct place to fix this is in `FilesystemOptions::find`, where we should propagate directory errors if and only if they occur at the top level search. This would incidentally also warn when we do workspace discovery, but this is probably still appropriate.

File: crates/uv-settings/src/lib.rs (hunk 1 above)
Comment (tyblue18) [reply crates/uv-settings/src/lib.rs]: I made this change, `from_directory` now returns a dedicated `Error::Directory` for both `uv.toml` and `pyproject.toml`. Also `find` emits a `warn!` (silent by default as suggested) when skipping these in parent directories.

File: crates/uv-workspace/src/workspace.rs (hunk 2 above)
Comment (tyblue18) [reply crates/uv-workspace/src/workspace.rs]: I reverted all `uv-workspace` changes, the fix now lives in settings discovery.

**Comment by tyblue18** [comment]
I reworked it as suggested: `FilesystemOptions::find` now propagates directory errors IF and only IF they occur at the top level of the search, and `!warn` otherwise. Since this runs during config resolution it will restore the error for commands that don't do project discovery (added a `uv venv` integration test for this), and this same handling also covers `uv.toml`. 
One edge case I wanted to confirm: a `pyproject.toml` directory in the CWD  when a valid project exists in a parent is still completely silent, workspace discovery succeeds and the settings search starts at the workspace root, so the CWD is never visited. That seems consistent with the " IFF at the top level search", but let me know if you want warning there too, and I can make the change.
